Initial Connection ID text

erickinnear opened the pull request.

ekinnear committed:
New frame, initial text around connection IDs

ekinnear committed:
Add new sections to better organize

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,82 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+The initial connection ID issued to the peer is sent by the endpoint as the
+Source Connection ID during the handshake, ensuring that the peer always begins
+the connection with at least one connection ID to use when sending.  When the
+endpoint receives the `requested_connection_ids` transport parameter during the
```
I think that we should defer addition of a `requested_connection_ids` transport parameter.  I'm not convinced that doing this is useful.  We should instead provide encouragement to make enough connection IDs available.  That number will be fairly stable.

MikeBishop replied:
We'll probably need a larger number when doing multipath, but we're not there yet.  Maybe something like:

```...to use when sending.  Endpoints SHOULD supply peers with additional connection IDs via NEW_CONNECTION_ID frames.  While each endpoint can choose how many connection IDs to issue, a recommended number of outstanding connection IDs is eight.```

The one case this doesn't really seem to cover is one over-eager endpoint (I'll give you 32!) and a constrained endpoint that doesn't want to track that many.  If it drops them, it risks running out; if it retires the excess, the peer will just flood it with more.

erickinnear replied:
Thanks, I pulled in this text

Re: over-eager endpoints, I think just dropping as many as you don't want to use is sufficient. 
i.e. if you give me 32 of them and I just save the first eight, I'm in exactly the same position as if you'd given me 8, and there's no cost on my end. The only extra cost is on the person who supplied 32.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,82 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+using a NEW_CONNECTION_ID frame to reduce the possibility of its peer running
+out of available connection IDs.
+
+It its peer selected non-zero-length connection IDs, an endpoint that receives a
```
"It its"?

erickinnear replied:
Fixed

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,82 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+or changes connection ID on an existing path, the packets will use different
+connection IDs in both directions. Likewise, when an endpoint receives a
+CONNECTION_ID_FINISHED frame, it SHOULD retire any connection IDs that were used
+in packets generated in response to that connection ID.
```
This paragraph is a little awkward.  The intent here is to say:

* when your peer changes, consider moving too
* when you stop using a connection ID, retire it

I thought that we had agreed that you don't have to use a new connection ID unless any part of the address tuple changes (for simplicity, both remote and local address and port, but in practice this is only really remote).

erickinnear replied:
Really good point. I've reworded this in 18cd68b, additional wording suggestions welcome!

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,82 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+
+If an endpoint's peer has selected a non-zero-length connection ID, the endpoint
+maintains a set of connection IDs received from the peer that it can use when
+sending packets.  The initial connection ID used when sending packets is sent by
```
This first sentence is duplicative of earlier text.

erickinnear replied:
We've got the same thing from two different perspectives to match the two sections of "issuer" and "consumer" of conn IDs. I removed the following sentence which was almost exactly the same text.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,82 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+sufficient connection IDs are available for future use, the endpoint
+SHOULD request additional connection IDs using the `requested_connection_ids`
+transport parameter.  The endpoint will then receive up to the requested number
+of connection IDs via NEW_CONNECTION_ID frames from its peer.
```
As before, I think that we should concentrate on having an ample supply and make recommendations about that.  Enough to migrate is what we should be aiming for.  @MikeBishop and I both arrived at a recommendation of 8, which covers enough for a few failed probes with a little padding to allow for the gap between when one is retired and a replacement arrives.  As low as 5 might work, but the margin for error seems a little tight.  The number will be higher if we get to multipath, but not a great deal (maybe +1 or +2 per expected active path).

erickinnear replied:
I like this, and agree with 8 as a reasonable recommendation.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,82 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+the peer to issue additional connection IDs via a NEW_CONNECTION_ID frame.
+
+Additionally, each connection ID MUST be used on packets sent from only one
+local address.  At any time, an endpoint MAY change to a new connection ID on a
```
This should be path.  Once a connection ID has been attached to a packet, that connection ID MUST NOT be sent on any other path.

ekr replied:
Well, MUST NOT be sent to any remote path. You might accidentally send it through a differeent path if you were mobile.

erickinnear replied:
This is the text that we had previously, but @ekr's point is what we need to be careful of -- there are cases where the client doesn't control which path it's being sent on. How do we want to word that?

martinthomson replied:
If you can't control the path, then it's the same as the case where you have a NAT rebinding that you are unaware of.  Don't worry about that case.  There are enough cases you can catch to make a careful prohibition useful though.

MikeBishop replied:
A NAT rebinding means that packets from the same local address look different when they arrive at the peer.  I think "local address" is fine; it's the language used in the connection migration section, so it's consistent.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,82 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+local address already in use.  An endpoint that migrates to a new local address
+may wish to retire all connection IDs that were used on the previous address
+using the CONNECTION_ID_FINISHED frame.  This allows the server to provide new
+connection IDs which are valid on any address until first use.
```
I don't understand this last sentence.  And why is this exclusively the server?

I think that you instead want to say that once a path is no longer in use, the endpoint SHOULD use the _FINISHED frame to retire those connection IDs.  (Don't bother mentioning that this will result in more connection IDs being provided.)

erickinnear replied:
Sorry, this shouldn't mention server specifically
We can just say should used _FINISHED to retire all ones from previous address.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -3139,14 +3191,12 @@ An endpoint sends a NEW_CONNECTION_ID frame (type=0x0b) to provide its peer with
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                          Sequence (i)                       ...
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 |   Length (8)  |          Connection ID (32..144)            ...
```
Not for this PR, but I was thinking that we might want to consider changing the format of this so that it supports having 1..16 connection IDs, all the same length.  Same for the _FINISHED.

erickinnear replied:
I was wondering the same thing, but kept it consistent with _NEW. 
I think it would be great to pack multiple into a single frame (in a different PR).

dtikhonov commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,82 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+
+An endpoint issues connection IDs to its peer for the peer to use when sending
+packets. This allows the endpoint control over the strategy used to interpret
+connection IDs marked on packets that it receives.  These connection IDs are
```
I don't think that connection IDs can be "marked on packets."  Maybe "connection IDs included in packets?"

MikeBishop replied:
Or simply "connection IDs of the packets"?

dtikhonov commented on draft-ietf-quic-transport.md:
```diff
@@ -1654,6 +1697,15 @@ disable_migration (0x0009):
 
+requested_connection_ids (0x000c):
+
+: A 32-bit unsigned integer value indicating the total number of connection IDs
```
A 32-bit value seems like an overkill.

erickinnear replied:
Leaning towards removing the TP entirely :)

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -1191,7 +1192,7 @@ different IP or port at either endpoint, due to NAT rebinding or mobility, as
 endpoint, as described in {{termination}}.
 
-## Connection ID
+## Connection ID {#connection-id}
```
This is actually unneeded in recent versions of kramdown-rfc2629 -- an anchorless section gets an implicit anchor from the name, which in this case would be... `#connection-id`

erickinnear replied:
Neat, thanks!

erickinnear replied:
Fixed in 3e789b8

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,82 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+When a receiver of a packet notices that its peer is now using a previously
+unused connection ID, it MAY choose to supply its peer with a new connection ID
+using a NEW_CONNECTION_ID frame to reduce the possibility of its peer running
+out of available connection IDs.
```
These paragraphs discuss the same thing -- suggested techniques to keep the peer from hitting bottom.  Maybe merge the paragraphs?

erickinnear replied:
Good thinking, fixed in 3e789b8

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,82 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+using a NEW_CONNECTION_ID frame to reduce the possibility of its peer running
+out of available connection IDs.
+
+It its peer selected non-zero-length connection IDs, an endpoint that receives a
```
"If the peer"; this still feels a little obtuse, and the synchronization problem remains.  I like @ekr's suggestion of requiring a shift when you see the peer change IP or port instead.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,82 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+or changes connection ID on an existing path, the packets will use different
+connection IDs in both directions. Likewise, when an endpoint receives a
+CONNECTION_ID_FINISHED frame, it SHOULD retire any connection IDs that were used
+in packets generated in response to that connection ID.
```
This is interesting, but risks the same synchronization problem.  Packets aren't necessarily "in response to" other packets in a QUIC sense.  Perhaps we turn it around and say "when the peer migrates to a new address, an endpoint SHOULD retire all CIDs which were used on packets directed to the previous address."  We should also retain the guidance that each CID only be used with a single remote address.

erickinnear replied:
Removed in response to 

Covering the rest of this in the summary below

MikeBishop reviewed (changes requested):
Generally good.  Two areas I think we need to fix, though:

- There are implicit synchronization issues here around responding to peer changes of CID.  We need some absolute reference to make synchronization feasible -- if it's not a sequence number, then perhaps remote address?  Otherwise, it becomes too easy to rotate CIDs into infinity.
- There's no discussion here of when Stateless Reset Tokens become invalid.  For simplicity, I'd suggest that an endpoint no longer recognizes the SRT associated with a given CID once it sends the _FINISHED frame.  A reasonable implementation might choose to wait an RTT to send _FINISHED after switching to catch any Stateless Resets that are coming back to it, but that doesn't need to be normative.

kazuho commented:
Now that sequence is removed, I think we need to specify how CIDs are deduplicated by the receiver of a NEW_CONNECTION_ID frame.

In my view, we have four states for CIDs:

* idle - a CID that has been received but not yet have been used nor retired
* active - a CID that is in use but not yet have been retired
* retiring - a CID that has been sent using a CONNECTION_ID_FINISHED frame (yet to be acknowledged)
* retired - a CID hat has been sent using a CONNECTION_ID_FINISHED frame that has been acknowledged

I think that we should specify that the receiver of a NEW_CONNECTION_ID frame must ignore contained CIDs that are either in idle, active, retiring state. I do not think that we want to require the receiver to check if the delivered CID matches one in retired state, because that requires unbounded amount of memory.

Having stated that, I wonder if we are actually free from race conditions that lead to unexpected connection teardown.

Consider the case where a client is starving for CIDs. As soon as it receives a NEW_CONNECTION_ID frame, it will send a probe using the CID that has been newly delivered. Probing can fail immediately, for example when there is some issue with the routing table of an alternate path. In such case, the client will be racing an ACK frame (for NEW_CONNECTION_ID) and a CONNECTION_ID_FINISHED frame.

If only the ACK gets lost, the server might retransmit the NEW_CONNECTION_ID frame and then immediately after that send an ACK for the CONNECTION_ID_FINISHED frame. That could lead to the ACK being delivered to the client before the retransmitted NEW_CONNECTION_ID frame if reordering happens.

To put it another way, there is a slight chance for an endpoint to receive a CID that is already in the retired state. In such case, the connection will be teared down by a stateless reset once the endpoint uses the CID.

Do we need to fix this issue? I think we have three options:
* ignore the issue, assuming that the chances of hitting the case is low
* change the definition of stateless reset from tearing down a connection to abandoning a CID, and also require the client to send a CONNECTION_ID_FINISHED frame specifying the CID that has been statelessly reset using another CID (if possible)
* add a time-wait state and require the receiver to retain retired CIDs for some time (e.g. 2SRTT) so that it could be used for deduplication

Personally, I think that we need to fix the issue in some way, because however the chances are rare, we might still hit the case considering the vast number of QUIC connections that we will have.

MikeBishop commented:
I agree.  Sequence was an easy way to solve this....  The idea of redefining Stateless Reset is interesting, but means you'd need as many round trips as you have active CIDs to actually confirm the connection has terminated.  At that point, SR doesn't seem noticeably better than just timing out.

I'm inclined to say that endpoints remember CIDs they've retired for a reasonable time, or until the _FINISHED is acknowledged, if they prefer.  Since the acknowledgement of _FINISHED already has to be tracked to know whether you need to retransmit it, that doesn't seem crippling, but 2SRTT should be enough for implementations that don't want to cross-connect acknowledgements to other state.

ekr commented on draft-ietf-quic-transport.md:
```diff
@@ -3192,6 +3234,48 @@ zero-length Destination Connection ID MUST treat receipt of a NEW_CONNECTION_ID
+
+: An 8-bit unsigned integer containing the length of the connection ID.  Values
+  less than 4 and greater than 18 are invalid and MUST be treated as a
+  connection error of type PROTOCOL_VIOLATION.
```
I don't think this is helpful. It's a PROTOCOL_VIOLATION to send an unknown CID here and so if you just enforce that, you will get the right behavior without a length check, as long as you check for 0.

martinthomson replied:
Retransmissions of CONNECTION_ID_FINISHED could arrive after the provider of connection IDs disposes of the old value.  That means that you can't generate errors unless you remember all previous connection IDs (this is the mirror-image of @kazuho's problem).  We can allow for PROTOCOL_VIOLATION if you like, but it might be hard to mandate.

ekr replied:
Well, the text below on line 3274 requires this.

martinthomson replied:
Yes, and that's a problem.

erickinnear replied:
Yeah, that’s a really good point, my initial feeling would be to not make it an error.

erickinnear replied:
Thoughts on keeping the length requirement and removing the "remember all the IDs" requirement?
i.e. we can't get rid of this text if we also get rid of 3274 and we need to get rid of 3274

MikeBishop replied:
Technically, you don't have to error on invalid CIDs.  You check whether the supplied CID is one you have associated with this connection; if so, remove it.  If the supplied CID is out-of-bounds, it won't be there.  If we feel we need an error case here, I think a MAY PV is sufficient.

erickinnear replied:
How about: 
```
An 8-bit unsigned integer containing the length of the connection ID.  Receipt of a value
   not previously issued to the peer as a valid connection ID MAY be treated as a
   connection error of type PROTOCOL_VIOLATION.
```

erickinnear replied:
Fixed in b54d0ce

ekr reviewed (commented):
I tend to think Mike's approach is best here.

kazuho commented:
@MikeBishop 
> I agree. Sequence was an easy way to solve this....

Honestly speaking I think I am fine with bringing back sequence. IMO, the key point of the @erickinnear's proposal is to explicitly signal every retirement of CIDs. Whether we use sequence numbers to track CIDs (by restoring the field in NEW_CONNECTION_ID frame and also by identifying the CID to be retired by sequence number) is a minor issue.

> The idea of redefining Stateless Reset is interesting, but means you'd need as many round trips as you have active CIDs to actually confirm the connection has terminated. At that point, SR doesn't seem noticeably better than just timing out.

That's a good point. I agree that it is not a good idea.

> I'm inclined to say that endpoints remember CIDs they've retired for a reasonable time, or until the _FINISHED is acknowledged, if they prefer. Since the acknowledgement of _FINISHED already has to be tracked to know whether you need to retransmit it, that doesn't seem crippling, but 2SRTT should be enough for implementations that don't want to cross-connect acknowledgements to other state.

SGTM.

erickinnear commented:
Re: sequence numbers, I think the other additional difference here beyond the _FINISHED communication is allowing all connection IDs to be in use at the same time. 

That probably requires removing the text around “when you see a new one starting to get used, issue another”, although that will help some of the synchronization. 

If you don’t need to advance in lockstep, then you don’t particularly need any ordering or index on connection ID either, it should be its own identifier (if that makes sense). 

For example (and I can reply up there to make the thread easier to keep track of), you no longer need the “idle” state at all (or at least not beyond the receiver, the point being you don’t need to keep it in sync across the network).

erickinnear commented:
I’d add another +1 to ekr about Mike’s comment - either you track the ack or you wait a reasonable time (and I think that was a good recommendation for time) and then throw it away. 

Re: A client getting a NEW_ frame that contains one that was just retired, if it’s held onto the CID for a couple of RTTs then we can assume it knows about it and will ignore it.

kazuho commented:
> I’d add another +1 to ekr about Mike’s comment - either you track the ack or you wait a reasonable time (and I think that was a good recommendation for time) and then throw it away.

I would suggest referring to the time only, because when reordering happens there is a chance for an ACK for the _FINISHED frame to arrive before the retransmitted NEW_CONNECTION_ID frame. See my comment above for the exact case.

ekinnear committed:
Editorial updates

ekinnear committed:
Reword coordinated switching paragraph

martinthomson commented:
In other cases, we have specified 3xRTO for cases where there is a tail.  We should probably use that constant here as well.  2sRTT is probably OK, but there's no reason to switch these off on a more aggressive timer.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1242,33 +1242,26 @@ Implementations SHOULD ensure that peers have the requested number of connection
 new connection ID for each connection ID retired with a CONNECTION_ID_FINISHED
-frame.
+frame. When a receiver of a packet notices that its peer is now using a
+previously unused connection ID, it MAY choose to supply its peer with a new
```
s/MAY/could - this isn't normative text

erickinnear replied:
Done, thanks

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -1242,33 +1242,26 @@ Implementations SHOULD ensure that peers have the requested number of connection
-of connection IDs via NEW_CONNECTION_ID frames from its peer.
+sending packets. To ensure that sufficient connection IDs are available for
+future use, the endpoint SHOULD request additional connection IDs using the
+`requested_connection_ids` transport parameter.  The endpoint will then receive
```
Are we dropping this transport parameter or not?

erickinnear replied:
Yup, looks like there's no objection to doing that, so I'm removing it. Just wanted to double check on that one.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,74 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+available connection IDs.  An implementation could do this by always supplying a
+new connection ID for each connection ID retired with a CONNECTION_ID_FINISHED
+frame. When a receiver of a packet notices that its peer is now using a
+previously unused connection ID, it MAY choose to supply its peer with a new
```
We don't really need a MAY for suggested strategies; "could" or "can" work here too.

erickinnear replied:
Changed to "can"

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,74 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+An endpoint that receives a packet with a previously unused connection ID,
+SHOULD also switch to sending with a different connection ID.  This can help to
+ensure that different connection IDs will be used in both directions when an
+endpoint migrates to a new path or changes connection ID on an existing path.
```
Again, I thought we had consensus on doing this on remote address change instead?

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -1207,41 +1208,74 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+
+All connection IDs issued by the peer are considered valid for use by the
+endpoint when sending packets until the connection ID is retired by the
+endpoint.  Endpoints may choose to stop using a given connection ID to send
```
s/may/can/

erickinnear replied:
Done

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -3192,6 +3226,48 @@ zero-length Destination Connection ID MUST treat receipt of a NEW_CONNECTION_ID
+
+An endpoint MUST treat receipt of a CONNECTION_ID_FINISHED frame containing a
+connection ID that it did not issue to its peer as a connection error of type
+PROTOCOL_VIOLATION.
```
As discussed above, remove this.

erickinnear replied:
Done

ekinnear committed:
Remove transport parameter and review comments

ekinnear committed:
Merge remote-tracking branch 'origin/master' into conn_id

erickinnear commented:
Removed transport parameter and updated with other comments. 
Added a note in the stateless reset section as well as one in the CONN_ID_FINISHED frame about invalidating stateless reset tokens. 

Remaining items that I'm aware of: 
1. Text around moving in response to someone else: 
```
An endpoint that receives a packet with a previously unused connection ID,
SHOULD also switch to sending with a different connection ID.  This can help to
ensure that different connection IDs will be used in both directions when an
endpoint migrates to a new path or changes connection ID on an existing path.
```
I'm not completely sure about recommending a change when some part of the 4-tuple changes, since a NAT rebinding would then cause you to "burn" a connection ID. That being said, I'm fine with making a change to something like: 
```
An endpoint that receives a packet with a different remote address than 
previously used SHOULD also switch to sending with a different connection
 ID.  This can help to ensure that different connection IDs will be used in both 
directions when an endpoint migrates to a new path or changes connection ID
 on an existing path.
```
That doesn't cover the case where one side just changed CIDs, do we still want to recommend that (i.e. either you see a different remote address or a new CID)? Seems kind of icky. 

2. Specifying 3xRTO for how long to remember a connection ID that was issued after sending FINISHED, something along these lines in "Consuming CIDs": 
```
An endpoint that retires a connection ID should retain knowledge of that
connection ID for a reasonable time after sending the CONNECTION_ID_FINISHED
frame, or until that frame is acknowledged.  A recommended time is three times
the current Retransmission Timeout (RTO) interval as defined in
{{QUIC-RECOVERY}}.  This prevents potential retransmissions of a
NEW_CONNECTION_ID frame from overlapping with the CONNECTION_ID_FINISHED frame
for the same connection ID.
```

Is there anything else here? I tried to capture everything from the conversations above.

MikeBishop commented:
Yes, it does cause you to burn a CID on rebind, but that's probably okay.  A rebind is most likely to appear after a period of quiescence, and we already recommend changing the CID you send with after you've been idle for a bit.  However, if we respond to CID changes, you run into one of two failure modes:  If the peer has a zero-length CID, you'll obviously never see them change.  If they have a non-zero-length CID, avoiding looping CID changes requires fuzzy heuristics.

However, if you change after a period of quiescence (when you *might* have been rebound) and the other side changes when you actually get rebound, the net effect is that you'll both change CIDs (if you're using them) on a rebind.

If you get a rebind mid-stream and you're using CIDs, you've already leaked the correlation -- doesn't matter whether the other side changes CID or not.  If it does, though, no harm done.  If you get a rebind mid-stream and you're not sending a CID, the connection will fail regardless, so who cares?

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -2578,8 +2606,11 @@ so.
+Tokens sent via NEW_CONNECTION_ID frames are invalidated when their associated
+connection ID is retired via a CONNECTION_ID_FINISHED frame
+({{frame-connection-id-finished}}).  This value is protected by encryption, so
+only client and server know this value.
```
With the intervening discussion of tokens and frames, it's no longer obvious what "This value" is.  I'd either swap the last two sentences or explicitly say "The token" again.  Probably with a preference toward the sentence-swap, since the invalidation of tokens is unrelated to them being issued under encryption.

erickinnear replied:
Sounds good, I'll swap them

erickinnear commented:
> However, if you change after a period of quiescence (when you might have been rebound) and the other side changes when you actually get rebound, the net effect is that you'll both change CIDs (if you're using them) on a rebind.

> If you get a rebind mid-stream and you're using CIDs, you've already leaked the correlation -- doesn't matter whether the other side changes CID or not. If it does, though, no harm done. If you get a rebind mid-stream and you're not sending a CID, the connection will fail regardless, so who cares?

Thanks! That's a really good point, I'll switch the text.

ekinnear committed:
Review comments

erickinnear commented:
Addressed all comments, anything I missed?

ekinnear committed:
Merge branch 'master' into conn_id

ekinnear committed:
One more review comment

kazuho commented on draft-ietf-quic-transport.md:
```diff
@@ -3219,6 +3249,47 @@ zero-length Destination Connection ID MUST treat receipt of a NEW_CONNECTION_ID
+
+: A connection ID of the specified length.
+
+An endpoint MUST NOT send this frame if it currently sending packets with a
```
"if it **is** currently sending"?

erickinnear replied:
Good catch! Added

ekinnear committed:
Move restriction out of length field

ekinnear committed:
Typo

kazuho commented on draft-ietf-quic-transport.md:
```diff
@@ -1224,41 +1225,77 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+frame. When a receiver of a packet notices that its peer is now using a
+previously unused connection ID, it can choose to supply its peer with a new
+connection ID using a NEW_CONNECTION_ID frame to reduce the possibility of its
+peer running out of available connection IDs.
```
> When a receiver of a packet notices that its peer is now using a previously unused connection ID, it can choose to supply its peer with a new connection ID using a NEW_CONNECTION_ID frame to reduce the possibility of its peer running out of available connection IDs.

I wonder if we need to allow this behavior.

Considering the fact that it is a MAY, endpoints cannot leverage from the optimization assuming that peers implement this.

I think it would be fine to just have a statement that suggests peers issuing new connection IDs as it receives _FINISHED frames.

erickinnear replied:
Fair enough. I'm not sure what an endpoint would do with the knowledge that the peer implements this, it's more of an opportunistic thing in my mind: Either the endpoint has no CIDs available when it needs one or it has a CID it can use. 

If you're writing code that issues CIDs, you might want to think about this -- it could help keep your peer from running out when they need it. There are a couple of other recommendations of this sort that are designed to help things go smoothly, but aren't strictly necessary if you think you know better when you're implementing things. 

Either way, I definitely agree that we need to suggest peers issue new IDs as it gets _FINISHED frames (and I think we do today, unless I'm mistaken).

kazuho replied:
> Either way, I definitely agree that we need to suggest peers issue new IDs as it gets _FINISHED frames (and I think we do today, unless I'm mistaken).

I agree that we suggest that, and my point is that we should only have that requirement.

The reason I prefer disallowing the peer to send additional CIDs before receiving _FINISHED frames is that such behavior requires the client to remember `max_probe_paths + 8` connection IDs. If we are to require that amount of memory to be reserved, why not raise the limit (i.e. 8)?

erickinnear replied:
Okay, thinking it through, I think I'm in agreement now. 
We don't ever require that a certain amount of memory is reserved on the person getting the NEW_ frames, they can happily ignore them and just work on a subset. The only potential problem there is if the issuer decides to retroactively lower the amount of CIDs that they maintain. 

So, if we remove this text and make the replacement on receipt of _FINISHED mandatory, then I think we're all set. 

And that keeps us on the same track with the "if you ever issue this many CIDs, you better be prepared to have that many CIDs you're dealing with forever" approach.

martinthomson replied:
I don't know if I agree though.  I would prefer to frame this differently:

1. you issue connection IDs as many as you feel like
2. you can't forget an issued connection ID unless your peer tells you they have forgotten it
3. you remember as many of the connection IDs you were given as you feel like and forget the rest
4. you can't reuse a connection ID for a different path, so forgetting it is a good idea
5. connection IDs you forget can be signaled with indicated with a FORGET_CONNECTION_ID frame if you need more or think you might need more

And then some advice:

* ideally a peer has as many connection IDs available to them as paths they might use, which is hard to predict, so allow for some spares (without multipath, 8 is probably plenty)
* as a peer tells you that is has forgotten connection IDs, it's best to issue them replacements so that they always have enough

erickinnear replied:
Generally, I'm completely on board with that. 

The only problem that would be solved by making sure that the other side *always* replaces when they get _FINISHED frames is this: 

1. Peer issues 16 CIDs
2. I only wanted 8, so I just ignore and throw out the rest
- Now normally this should work fine, we'd just all keep on going with 8 -
3. *If* the peer decides that they only want to maintain 8 and they just gave me 8 "extra" to start with, then I'm screwed if
4. I retire all 8 CIDs (or 7, or whatever)
5. The peer never sends me new ones

Now I'm stuck because they essentially retracted their giving of 16. 

We can solve all of that if we require replacement when _FINISHED.

Other alternative: Add a flag to _FINISHED saying "do/do not replace" and in step (2) above, I *tell* the peer that I'm throwing away 8. 
Not a huge fan of that, though. 

Or, and I'd be okay with this, we just say "the recommendation is there for a reason" and leave as is in the current text. Technically a peer could do that wrong, but they could also just not issue a lot of CIDs too, since 8 is only a recommendation, so we don't need to try to plug this hole and not the others.

martinthomson replied:
I think that this risk is easier acknowledged, but not addressed.  Yes, if you throw away half of the connection IDs, there's a chance that you will need them.  You took a risk and you might need to make a new connection to repair it.  That's not terrible.

erickinnear replied:
Yeah I like that perspective, let's leave this text for now

kazuho replied:
I think I might agree to what @martinthomson says, assuming that we will not introduce a transport parameter that tells the peer how many CIDs the endpoint can remember.

kazuho commented on draft-ietf-quic-transport.md:
```diff
@@ -1224,41 +1225,77 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+connection ID using a NEW_CONNECTION_ID frame to reduce the possibility of its
+peer running out of available connection IDs.
+
+An endpoint that receives a packet with a different remote address than
```
"with a different remote address **or with a different CID** than previously used"?

erickinnear replied:
We had that before, do we want to go back to that (now with remote address too)?

kazuho replied:
I think theoretically we should require the peer to switch to a new CID on an existing 5-tuple when an endpoint switches to a new one.

Such behavior is beneficial to protect the lifetimes of connections between servers. It is natural to expect many QUIC connections between servers to share a single 5-tuple. In such case, an endpoint might want to protect the lifetime of each connection by randomly switching to a new CID. But unless we require CID migration on an existing path to be complementary, you cannot protect it...

erickinnear replied:
So we have: 
> An endpoint that receives a packet with a different remote address or with a different connection ID than previously used SHOULD also switch to sending with a different connection ID.

I don't think it can be stronger than a SHOULD because the other endpoint may not have enough CIDs to actually switch.

martinthomson replied:
More to the point, if you require a change, you create the potential for infinite regress.  "They changed, so I will change" can repeat until you run out of connection IDs.

erickinnear replied:
Agreed. 
As to keeping "or with a different connection ID" in or not, let's continue that once we've posted this to the main repo, seems minor enough.

martinthomson replied:
Oh, and one model that was suggested was greasing of connection ID use.  I could send with another connection ID 1 out of N packets.  That could trigger exhaustion as well.

MikeBishop replied:
If you switch CIDs without changing addresses they're already linked to any observers, so it doesn't matter whether the remote side changes.  If you switch addresses without changing CIDs, they're already linked to any observers, so it doesn't matter whether the remote side changes.

The case we actually *need* to cover is ensuring that, if an endpoint changes address and CID at the same time, the peer absolutely does not respond with the old CID.  Any rule that produces this is sufficient, even if it also produces some gratuitous changes along the way.

I tend to think switching on CID change is simply not workable or desirable, because it's hard to avoid looping.  It also prevents greasing, as mentioned, which is nice to keep but not necessarily essential.  Switching on endpoint address covers the scenario we need.

MikeBishop commented:
I think this is in good enough shape to broaden the discussion.  @erickinnear, can we turn this into a PR on the base repo, or would you like me or @martinthomson to do it?

erickinnear commented:
I'll submit this shortly as a PR against the base repo, thanks for all the help here!

ekinnear committed:
Minor wording

ekinnear committed:
Merge remote-tracking branch 'origin/master' into conn_id
